### PR TITLE
Add feature for setting docs & protocolHandlers from constructor.

### PR DESCRIPTION
**Pull request by aljones15**
Adds a new minor feature that allows you to pass in a documents Map and a protocolHandler map to the constructor avoiding the loop of using `addStatic` and `setProtocolHandler`.
This makes creation of multiple documentLoaders much easier.
Also adds 2 tests to the project that use this constructor.

### CODE REVIEW

File: lib/JsonLdDocumentLoader.js [hunk 1]
```diff
@@ -1,11 +1,27 @@
...
  */
-
 export class JsonLdDocumentLoader {
-  constructor() {
-    this.documents = new Map();
-    this.protocolHandlers = new Map();
+  /*
+   * @param {object} options - Document Loader Options.
+   * @param {Map<string, object>} [options.documents = new Map()] - A Map
+   * containing urls and their associated contexts.
```
Comment (davidlehn) [inline lib/JsonLdDocumentLoader.js]:
```suggestion
   * containing URLs and their associated contexts.
```

File: lib/JsonLdDocumentLoader.js [hunk 2]
```diff
@@ -1,11 +1,27 @@
...
+   * @param {Map<string, {get: Function}>} [options.protocolHandlers =
+   *   new Map()] - A map containing protocolHandlers for the loader.
+  */
```
Comment (davidlehn) [inline lib/JsonLdDocumentLoader.js]:
```suggestion
   */
```

File: lib/JsonLdDocumentLoader.js [hunk 3]
```diff
@@ -1,11 +1,27 @@
...
 export class JsonLdDocumentLoader {
-  constructor() {
-    this.documents = new Map();
-    this.protocolHandlers = new Map();
+  /*
+   * @param {object} options - Document Loader Options.
+   * @param {Map<string, object>} [options.documents = new Map()] - A Map
+   * containing urls and their associated contexts.
+   * @param {Map<string, {get: Function}>} [options.protocolHandlers =
+   *   new Map()] - A map containing protocolHandlers for the loader.
```
Comment (davidlehn) [inline lib/JsonLdDocumentLoader.js]:
```suggestion
   *   new Map()] - A Map containing protocolHandlers for the loader.
```

**Review by davidlehn — COMMENTED** [review]
- Please add some docs/examples to the README for this feature.
- It might also be worth noting these are mutable params and direct changes or changes via the other API calls will effect every loader using the Maps.
- I do have a concern that this is exposing internal implementation details as an API.  I'm not sure what the future ramifications of that might be, if any.

**Comment by dlongley** [comment]
> I do have a concern that this is exposing internal implementation details as an API. I'm not sure what the future ramifications of that might be, if any.

Yeah, I agree with this. Perhaps it would be better to allow `addStatic` to take a `Map` of documents that will be processed and added to the internal implementation instead of passing and using something directly in the constructor. It could get weird with the map being used by reference and potentially updated later as well. I don't know that we have the same kind of use case for `setProtocolHandler` where adding the ability to accept a `Map` adds much, so we could probably skip that -- unless I'm missing something there.

**Comment by aljones15** [comment]
> > I do have a concern that this is exposing internal implementation details as an API. I'm not sure what the future ramifications of that might be, if any.
> 
> Yeah, I agree with this. Perhaps it would be better to allow `addStatic` to take a `Map` of documents that will be processed and added to the internal implementation instead of passing and using something directly in the constructor. It could get weird with the map being used by reference and potentially updated later as well. I don't know that we have the same kind of use case for `setProtocolHandler` where adding the ability to accept a `Map` adds much, so we could probably skip that -- unless I'm missing something there.

you do have a point, the Map is by reference and could be mutated elsewhere. Perhaps we could simply copy the Maps passed in? As protocolHandlers it's set in the constructor so I just included it there.

did this here: https://github.com/digitalbazaar/jsonld-document-loader/pull/11/commits/7a4bf3ebf9504b8485510d42f139a7e680a64a22

Example:

```js
const mapOne = new Map([['one', 1]]);
const mapTwo = new Map(mapOne);
mapOne.set('one', 2);
mapOne.get('one');
// returns 2
mapTwo.get('one');
// returns 1
```

I can add a test to ensure mutations to a Map passed into the documentLoader do not mutate entries in documents if wanted.

**Comment by aljones15** [comment]
@davidlehn @dlongley also of note: the existing `addStatic` function copies the document by reference into `this.documents` which means mutations to the original document will show up in `this.documents` so maybe we need to clone documents on add?

**Comment by aljones15** [comment]
closing as this does not appear to be the direction this library wants to go in.

**Pull request closed without merging**